docs/madr/decisions/031-automatic-rechable-services.md

```diff
@@ -0,0 +1,263 @@
+1. `Mesh/Mesh` - this won't help us because that's the default, anyone can talk to anyone and we can't
+
+2. `MeshService/Mesh`  
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: MeshTrafficPermission
+metadata:
+  namespace: kuma-system
+  name: mtp-ms-to-mesh
+spec:
+  targetRef:
+    kind: MeshService
+    name: a
+  from:
+    - targetRef:
+        kind: Mesh
+      default:
+        action: Allow
+```
+
+```mermaid
+graph TD
+    B-->A
+    C-->A
+    D-->A
+```
+
+We could either define an edge between every service to that service or have a special list that contains allIncomingAllowed.
+Services from that list would always be added as an outgoing edge.
+
+- `Mesh/MeshService`
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: MeshTrafficPermission
+metadata:
+  namespace: kuma-system
+  name: mtp-mesh-to-ms
+spec:
+  targetRef:
+    kind: Mesh
+  from:
+    - targetRef:
+        kind: MeshService
+        name: a
+      default:
+        action: Allow
+```
+
+```mermaid
+graph TD
+    A-->B
+    A-->C
+    A-->D
+```
+
+This is simply `A` can talk to anyone, this could be either adding and outgoing edge from this service to every other service or having it on a list called allOutgoing and not bothering to trim the configuration.
+
+3. `MeshService/MeshService` - already talked about, this is the easy case
+
+4. `MeshServiceSubset` 
+
+Unfortunately there is one option here,
+we need to treat `MeshServiceSubset` just like `MeshService`.
+This is because we can't trim individual endpoints because the data is not available at that point, and
+we have to accept the fact that we will miss out on (small) parts of the config that could be trimmed.
+
+5. `MeshSubset`
+
+This one is a bit more tricky because the potential radious is a bit bigger.
+The approach has to be the same as in `MeshServiceSubset`, meaning: treat `MeshSubset` as just `Mesh`.
+
+There is a case that makes sense using `k8s.kuma.io/namespace: xyz` to allow only traffic from a specific namespace,
+so this needs to be documented that it won't improve the performance.
+
+#### Migration
+
+A typical migration to Mesh involves bringing services gradually into the Mesh.
+If the system is small enough, after several iterations, it will be enabled for the whole system (every service).
+
+If the system is too big to do this we need to do extra work in the iterations (defining MeshTrafficPermissions along the way).
+
+1. 0 services are in the Mesh
+2. Bring X services into the Mesh (where X is small enough, so the Mesh can handle it)
+3. Enable MeshTrafficPermission for services brought into the mesh (the performance improves because we trimmed the configs)
+4. Repeat point 2-3 until all services are in the Mesh
+
+The good thing about how this migration is set up is that we can gradually do it without any interruption to the system.
+The downside of it is that **a service owner needs to know EXACTLY which services consume their service** to avoid any disruptions.
```

Comment:
No it doesn't. It's an optimisation the problem we're trying to address is that we have an algorithm that scales in O(n^2) and must reduce the complexity.
For any service from which we define a restrictive MTP we reduce the complexity. If it's perfect then you go to the optimal solution. If it's not perfect you are still better than worst case.

Think about it just telling people to start with:

```yaml
type: MTP
name: namespace-controlled
spec:
  targetRef:
    kind: MeshSubset
    tags:
       namespace: my-ns
  from:
    - kind: MeshSubset
      tags:
        namespace: my-ns
      action: Allow
```

Just do this for each of your namespace and then selectively allow services with:

```yaml
type: MTP
name: ns-allow
spec:
   targetRef:
      kind: MeshService
      name: my-public
   from:
      - targetRef:
            kind: Mesh
         action: Allow
```

Scale this to 10 namespaces with 10 services per namespace with 2 services per ns accessible outside their own service.

Then you have per instance 10 + (9 * 2) = 28 services to know about vs 100 services in our current naive case this is a massive improvement and without much sacrifice.

Even if you take a super open mesh with 1k services who only have 3 services with very restrictive MTP the number goes down quickly 1000^2 = 1M vs  997^2 + (the few services that do actually access the 3 restricted services)

Reply:
I agree but if I remember correctly my chat with @jakubdyszkiewicz we had a problem with `subsets` - https://github.com/kumahq/kuma/pull/8084/files#diff-28d6e9df351b905115d7d3e1c54ba2075abd0f57c370d093917530af05935e75R159 . Maybe @jakubdyszkiewicz figured out a way around this.

Reply:
We support any subset in from and subsets with only predefined tags in top level target ref. Those predefined tags are
`k8s.kuma.io/namespace`, `k8s.kuma.io/service` and `k8s.kuma.io/port`

Reply:
Added this scaling example and specific labels in `Subsets` section.